src/java/org/apache/cassandra/service/consensus/migration/ConsensusMigrationMutationHelper.java

```diff
@@ -0,0 +1,399 @@
+                }
+                catch (RetryOnDifferentSystemException | CoordinatorBehindException e)
+                {
+                    throw e;
+                }
+                catch (Exception e)
+                {
+                    failure = Throwables.merge(failure, e);
+                }
+
+                if (failure != null)
+                    throw unchecked(failure);
+            }
+            catch (RetryOnDifferentSystemException e)
+            {
+                writeMetrics.mutationRetriedOnDifferentSystem.mark();
+                writeMetricsForLevel(consistencyLevel).mutationRetriedOnDifferentSystem.mark();
+                logger.debug("Retrying mutations on different system because some mutations were misrouted");
+                continue;
+            }
+            catch (CoordinatorBehindException e)
+            {
+                logger.debug("Retrying mutations now that coordinator has caught up to cluster metadata");
+                continue;
+            }
+            break;
+        }
+    }
+
+    public static void validateSafeToExecuteNonTransactionally(IMutation mutation) throws RetryOnDifferentSystemException
+    {
+        if (mutation.allowPotentialTransactionConflicts())
+            return;
+
+        // System keyspaces are never managed by Accord
+        if (SchemaConstants.isSystemKeyspace(mutation.getKeyspaceName()))
+            return;
+
+        ClusterMetadata cm = ClusterMetadata.current();
+
+        DecoratedKey dk = mutation.key();
+        // Check all the partition updates and if any can't be done return an error response
+        // and the coordinator can retry with things correctly routed
+        boolean throwRetryOnDifferentSystem = false;
+        // Track CFS so we only mark each one once
+        Set<TableId> markedColumnFamilies = null;
+        for (PartitionUpdate pu : mutation.getPartitionUpdates())
+        {
+            ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(pu.metadata().id);
+            TableMetadata tm = cfs.metadata();
+            if (tokenShouldBeWrittenThroughAccord(cm, tm, dk.getToken()))
+            {
+                throwRetryOnDifferentSystem = true;
+                if (markedColumnFamilies == null)
+                    markedColumnFamilies = new HashSet<>();
+                if (markedColumnFamilies.add(cfs.getTableId()))
+                    cfs.metric.mutationsRejectedOnWrongSystem.mark();
+                logger.debug("Rejecting mutation on wrong system to table {}.{}", cfs.keyspace.getName(), cfs.name);
+                Tracing.trace("Rejecting mutation on wrong system to table {}.{} token {}", cfs.keyspace.getName(), cfs.name, dk.getToken());
+            }
+        }
+        if (throwRetryOnDifferentSystem)
+            throw new RetryOnDifferentSystemException();
+    }
+
+    public static boolean tokenShouldBeWrittenThroughAccord(@Nonnull ClusterMetadata cm, @Nonnull TableMetadata tm, @Nonnull Token token)
```

Comment:
We should be passing in the `TableId` and using it to retrieve the table metadata from the provided `ClusterMetadata`. Passing in table metadata directly like this makes it possible to read table params from a different epoch than the cm passed in, which I think is possible in invocation in `BlockingReadRepair`

Reply:
That forces redundant map lookups if it is the approach we take everywhere. If we went to the trouble of looking up the `TableMetadata` shouldn't we keep using it?

Reply:
Not if it can break correctness like it can here, that part definitely needs to be fixed. Unless we can assert that the TableMetadata object is from the same epoch as the ClusterMetadata object, we shouldn't be doing this.

Reply:
OK, WFM. Will change it and keep an eye out for similar instances.